cpp/arcticdb/python/python_handlers.cpp

```diff
@@ -20,67 +22,42 @@ namespace arcticdb {
 /// When numpy creates an empty array its type is float64. We want to mimic this because:
 /// i) There is no equivalent to empty value
 /// ii) We want input dataframes to be exact match of the output and that includes the type
-[[nodiscard]] static inline py::dtype generate_python_dtype(const TypeDescriptor &td, stride_t type_byte_size) {
-    if (is_empty_type(td.data_type())) {
-        return py::dtype{"f8"};
-    }
-    return py::dtype{fmt::format("{}{:d}", get_dtype_specifier(td.data_type()), type_byte_size)};
-}
 
-/// @important This calls pybind's initialize array function which is NOT thread safe. Moreover, numpy arrays can
-/// be created only by the thread holding the GIL. In practice we can get away with allocating arrays only from
-/// a single thread (even if it's not the one holding the GIL). This, however, is not guaranteed to work.
-/// @todo Allocate numpy arrays only from the thread holding the GIL
-[[nodiscard]] static inline PyObject* initialize_array(
-    const pybind11::dtype &descr,
-    const shape_t shapes,
-    const stride_t strides,
-    const void *source_ptr,
-    std::shared_ptr<Column> owner,
-    std::mutex &creation_mutex
-) {
-    std::lock_guard creation_guard{creation_mutex};
-    // TODO: Py capsule can take only void ptr as input. We need a better way to handle destruction
-    //  Allocating shared ptr on the heap is sad.
-    auto *object = new std::shared_ptr<Column>(std::move(owner));
-    auto arr = py::array(descr, {shapes}, {strides}, source_ptr, py::capsule(object, [](void *obj) {
-        delete reinterpret_cast<std::shared_ptr<Column> *>(obj);
-    }));
-    return arr.release().ptr();
+
+static inline PyObject** fill_with_none(PyObject** ptr_dest, size_t count, SpinLock& spin_lock) {
+    auto none = py::none();
+    for(auto i = 0U; i < count; ++i)
+        *ptr_dest++ = none.ptr();
+
+    spin_lock.lock();
```

Comment:
Should we check we're holding the GIL here?

Reply:
I think we'll get away without it with none, as it is a global singleton, so there will never be any allocation